src/world/Actor/Chara.cpp

```diff
@@ -551,28 +517,44 @@ void Sapphire::Entity::Chara::addStatusEffect( StatusEffect::StatusEffectPtr pEf
   status.index = static_cast< uint8_t >( nextSlot );
   status.param = pEffect->getParam();
 
-  sendToInRangeSet( statusEffectAdd, isPlayer() );
+  float totalShieldValue = 0;
+  for( auto effectIt : m_statusEffectMap )
+  {
+    auto statusEffect = effectIt.second;
+    if( static_cast< Common::StatusEffectType >( statusEffect->getEffectEntry().effectType ) == Common::StatusEffectType::Shield )
+    {
+      totalShieldValue += statusEffect->getEffectEntry().effectValue1;
+    }
+  }
+
+  if( totalShieldValue > 0 )
+  {
+    totalShieldValue /= getMaxHp();
+    totalShieldValue *= 100;
+    statusEffectAdd->data().shieldPercentage = totalShieldValue >= 255 ? 255 : static_cast< uint8_t >( totalShieldValue );
+  }
+
+  sendToInRangeSet( statusEffectAdd, true );
+  sendStatusEffectUpdate(); // although client buff displays correctly without this but retail sends it so we do it as well
 }
 
-/*! \param StatusEffectPtr to be applied to the actor */
-void Sapphire::Entity::Chara::addStatusEffectById( uint32_t id, int32_t duration, Entity::Chara& source, uint16_t param )
+void Sapphire::Entity::Chara::addStatusEffectById( uint32_t id, int32_t duration, Entity::Chara& source, uint16_t param, uint64_t lastTickOverride )
 {
   auto effect = StatusEffect::make_StatusEffect( id, source.getAsChara(), getAsChara(), duration, 3000, m_pFw );
   effect->setParam( param );
+  effect->setLastTick( lastTickOverride == 0 ? Util::getTimeMs() : lastTickOverride );
   addStatusEffect( effect );
 }
 
-/*! \param StatusEffectPtr to be applied to the actor */
-void Sapphire::Entity::Chara::addStatusEffectByIdIfNotExist( uint32_t id, int32_t duration, Entity::Chara& source,
-                                                             uint16_t param )
+void Sapphire::Entity::Chara::addStatusEffectByIdIfNotExist( uint32_t id, int32_t duration, Entity::Chara& source, uint16_t param, uint64_t lastTickOverride )
```

Comment:
why is the last tick overridden? if you're refreshing a status you should just update the timer on the existing status and resend that instead of constructing new objects everywhere

Reply:
it needs to recalculate damage/heal, reset uses, reapply shield etc, it's not a timer extend

Reply:
ah, I missed that, reasonable but it would be better to have something to 'reset' a status rather than having to nuke it entirely and recreate it.

Reply:
I think just create a new one looks cleaner though

Reply:
Maybe, but doing something like `status->refresh(...)` which in effect creates a 'new' status but still preserves needed data seems like a much cleaner approach to me

Reply:
you don't have to alloc any more memory to do that either so it's a bit faster as well

Reply:
or is it? you are going to modify the target status map in the status itself, in most existing access to the map it's in a iteration and can't really modify the map anyway. there's a markToRemove flag I created so status itself don't have to care about those things.

Reply:
you're not modifying the map if you refresh/modify the status directly, it's just a ptr whereas your add/remove does modify the map, though that's not really a concern of mine - it doesn't really matter.

Reply:
oops I need to go back to school and learn English

Reply:
lmao you're fine, doesn't help that half the stuff I say makes zero sense. let me know if something isn't clear or whatever and I can try explaining it in a not retarded way